usr/lib/common/mech_openssl.c

```diff
@@ -5566,4 +5566,299 @@ CK_RV openssl_specific_ibm_dilithium_generate_keypair(STDLL_TokData_t *tokdata,
+    if (bld == NULL) {
+        TRACE_ERROR("OSSL_PARAM_BLD_new failed\n");
+        rc = CKR_HOST_MEMORY;
+        goto out;
+    }
+
+    if (private_key) {
+        if (OSSL_PARAM_BLD_push_octet_string(bld, OSSL_PKEY_PARAM_PRIV_KEY,
+                                             priv_key, priv_len) != 1) {
+            TRACE_ERROR("OSSL_PARAM_BLD_push_octet_string failed\n");
+            rc = CKR_FUNCTION_FAILED;
+            goto out;
+        }
+    }
+
+    if (OSSL_PARAM_BLD_push_octet_string(bld, OSSL_PKEY_PARAM_PUB_KEY,
+                                         pub_key, pub_len) != 1) {
+        TRACE_ERROR("OSSL_PARAM_BLD_push_octet_string failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    params = OSSL_PARAM_BLD_to_param(bld);
+    if (params == NULL) {
+        TRACE_ERROR("OSSL_PARAM_BLD_to_param failed\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    pctx = EVP_PKEY_CTX_new_from_name(NULL, alg_name, NULL);
+    if (pctx == NULL) {
+        TRACE_ERROR("EVP_PKEY_CTX_new_from_name failed for '%s'\n", alg_name);
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    if (EVP_PKEY_fromdata_init(pctx) != 1) {
+        TRACE_ERROR("EVP_PKEY_fromdata_init failed for '%s'\n", alg_name);
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    if (EVP_PKEY_fromdata(pctx, pkey, private_key ? EVP_PKEY_KEYPAIR :
+                                                    EVP_PKEY_PUBLIC_KEY,
+                          params) != 1) {
+        TRACE_ERROR("EVP_PKEY_fromdata failed for '%s'\n", alg_name);
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+out:
+    if (priv_key != NULL) {
+        OPENSSL_cleanse(priv_key, priv_len);
+        free(priv_key);
+    }
+    if (pub_key != NULL)
+        free(pub_key);
+    if (pctx != NULL)
+        EVP_PKEY_CTX_free(pctx);
+    if (bld != NULL)
+        OSSL_PARAM_BLD_free(bld);
+    if (params != NULL)
+        OSSL_PARAM_free(params);
+
+    return rc;
+}
+
+CK_RV openssl_specific_ibm_dilithium_sign(STDLL_TokData_t *tokdata,
+                                          SESSION *sess,
+                                          CK_BBOOL length_only,
+                                          const struct pqc_oid *oid,
+                                          CK_BYTE *in_data,
+                                          CK_ULONG in_data_len,
+                                          CK_BYTE *signature,
+                                          CK_ULONG *signature_len,
+                                          OBJECT *key_obj)
+{
+    struct openssl_ex_data *ex_data = NULL;
+    EVP_PKEY *pkey = NULL;
+    CK_RV rc = CKR_OK;
+    EVP_PKEY_CTX *ctx = NULL;
+    const char *alg_name;
+    size_t siglen;
+
+    UNUSED(tokdata);
+    UNUSED(sess);
+
+    alg_name = openssl_get_pqc_oid_name(oid);
+    if (alg_name == NULL) {
+        TRACE_ERROR("Dilithium key form is not supported by oqsprovider\n");
```

Comment:
Hmm, why not include the oid into the trace msg

Reply:
The oid is a byte array, hard to print....